### PR TITLE
use threadpool and tornado queue for messaging

### PR DESCRIPTION
**Pull request by minrk**
- threadpool limits concurrent builds and thus DOS by spawning runaway threads
- tornado Queue removes sleep when waiting for events

### CODE REVIEW

**Comment by choldgraf** [comment]
Could you give a short description of what this adds? Or if @yuvipanda thinks the code looks good I'm cool w/ it too...I probably don't have the skillset to properly review this :-/

File: binderhub/build.py [hunk 1]
```diff
@@ -57,7 +55,7 @@ def get_cmd(self):
 
     def progress(self, kind, obj):
         """Put the current action item into the queue for execution."""
-        self.q.put_nowait({'kind': kind, 'payload': obj})
+        IOLoop.instance().add_callback(self.q.put, {'kind': kind, 'payload': obj})
```
Comment (yuvipanda) [inline binderhub/build.py]: I'm trying to wrap my head around this. This looks like we're wanting to call on the main thread a call to put. However, put can block, and put_nowait is just put with block=False set. It's also safe to put things into queue from multiple threads. Why this change?

File: binderhub/builder.py [hunk 2]
```diff
@@ -152,55 +152,57 @@ def get(self, provider_prefix, spec):
...
-
-        while True:
-            # FIXME: use Futures
-            try:
-                progress = q.get_nowait()
-            except Empty:
-                yield gen.sleep(0.5)
-                continue
+        while not done:
+            progress = yield q.get()
```
Comment (yuvipanda) [inline binderhub/builder.py]: Does this work? I don't see anything about generator / async support in https://docs.python.org/3/library/queue.html

File: binderhub/build.py (hunk 1 above)
Comment (minrk) [reply binderhub/build.py]: This is not a stdlib `queue.Queue`, it's a tornado `queues.Queue`.

`progress` is called from a thread and `q.put` must be called from the main thread. The only threadsafe way to call these things is to place them on the main thread via `.add_callback()`.

File: binderhub/builder.py (hunk 2 above)
Comment (minrk) [reply binderhub/builder.py]: This is not a stdlib `queue.Queue`, it's a tornado `queues.Queue`, which is like a Queue except it's async (hence `yield`).

**Comment by minrk** [comment]
> Could you give a short description of what this adds?

This uses a ThreadPool to limit concurrent builds and remove the spawning of new threads for every build. As a result, there is now a limit on the number of concurrent builds. It also switches to using the tornado async Queue instead of the stdlib Queue for communication. tornado Queues are async, which removes the need for a `gen.sleep` because we can just wait for messages on the Queue with `yield q.get()`.

**Comment by choldgraf** [comment]
Sounds good - if we hit the limit of concurrent builds, what happens? Does it just hang until a spot is opened up? Or error?

**Comment by minrk** [comment]
It'll wait

File: binderhub/build.py (hunk 1 above)
Comment (yuvipanda) [reply binderhub/build.py]: Aaah, I totally missed that change! This all makes sense now.
